stablehlo/reference/ProcessGrid.h

```diff
@@ -0,0 +1,127 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#ifndef STABLEHLO_REFERENCE_PROCESSGRID_H
+#define STABLEHLO_REFERENCE_PROCESSGRID_H
+
+#include <condition_variable>
+#include <cstdint>
+#include <map>
+#include <mutex>
+#include <utility>
+
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/Support/LLVM.h"
+#include "stablehlo/reference/Tensor.h"
+
+namespace mlir {
+namespace stablehlo {
+
+using ChannelId = int64_t;
+
+// StableHLO `process_id`.
+struct ProcessId {
+  /// StableHLO `replica_id`.
+  uint32_t replicaId;
+
+  /// StableHLO `partition_id`.
+  uint32_t partitionId;
+
+  bool operator<(const ProcessId &other) const;
+
+  bool operator==(const ProcessId &other) const;
+
+  bool operator!=(const ProcessId &other) const;
+};
+
+// StableHLO `process_group`.
+struct ProcessGroup : public SmallVector<ProcessId> {};
+
+// StableHLO `process_groups`.
+struct ProcessGroups : public SmallVector<ProcessGroup> {};
+
+class RendezvousResult {
+ public:
+  /// Iterates through the map `result_` and returns the value associated with
+  /// the key `processId`.
+  /// If key is not found, return an empty `Tensor`.
+  Tensor lookup(ProcessId processId);
+
+  /// Inserts `tensor` into the map `result_` using the key `processId`.
+  void insert(ProcessId processId, Tensor tensor);
+
+  /// Erases all elements in the map `result_`.
+  void clear();
+
+  /// Returns the size of `result_`.
+  size_t size();
+
+ private:
+  ///
+  std::map<ProcessId, Tensor> result_;
+};
+
+/// Class to model a process grid.
+class ProcessGrid {
+ public:
+  /// \name Constructors
+  /// @{
+  ProcessGrid(uint32_t numReplicas, uint32_t numPartitions);
+  /// @}
+
+  /// StableHLO `cross_replcia`.
+  ProcessGroups crossReplica(SmallVector<SmallVector<uint32_t>> replicaGroups);
+
+  /// StableHLO `cross_partition`.
+  ProcessGroups crossPartition(
+      SmallVector<SmallVector<uint32_t>> partitionGroups);
+
+  /// Each participating process in the `processGroup` appends its data
+  /// `operand` to the `channels_` map using the pair (processGroup, channelId)
+  /// as a key. `channels_` is cleared before any process adds its data.
+  /// `rendezvous` then returns `RendezvousResult` containing a mapping of
+  /// `processId` to its data `operand` once all participating processes have
+  /// successfully added their data. Throws an error after a timeout when
+  /// synchronization deadlocks.
```

Comment:
Same as above - let's write this comment in terms of a public API.

Reply:
The current wording is still formulated in terms of the implementation details which involve an internal map with an internal key - something that users of this API shouldn't need to know about.

```C++
  /// Synchronize a StableHLO process with the `processId` with other StableHLO
  /// processes in the `processGroup` using a `channelId`.
  ///
  /// A call to this method represents a barrier, i.e. it blocks the calling
  /// OS thread until all StableHLO processes from the `processGroup` call this 
  /// method with the same `channelId`. If the calling OS thread doesn't 
  /// correspond to the StableHLO process with `processId`, the behavior is 
  /// undefined.
  ///
  /// If any of the StableHLO processes from `processGroup` fails to arrive
  /// at the barrier within 3 seconds, the `rendezvous` fails with a fatal 
  /// error. This is to make sure that errors in underlying StableHLO programs 
  /// or bugs in the StableHLO interpreter don't deadlock the entire 
  /// interpreter.
  ///
  /// At the barrier, each StableHLO process contributes a tensor, and these
  /// tensors are accumulated in `RendezvousResult` which is returned to all
  /// callers once the barrier has been reached by all StableHLO processes.
```